cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -1130,151 +1217,198 @@ aggregate_reader_metadata::select_row_groups(
+            rows_to_read,
+            {},
+            std::vector<size_t>(per_file_metadata.size(), 0),
+            total_row_groups,
+            {num_filtered_row_groups, num_filtered_row_groups}};
+  }
+
+  // Vector to store all row group indices across all sources
+  std::vector<std::vector<size_type>> all_row_group_indices;
+
+  // Span to track the current row group indices
+  host_span<std::vector<size_type> const> current_row_group_indices;
+
+  // If input row group indices are not specified, populate the vector of all row group indices
+  if (row_group_indices.empty()) {
+    all_row_group_indices.reserve(per_file_metadata.size());
+    std::transform(per_file_metadata.cbegin(),
+                   per_file_metadata.cend(),
+                   std::back_inserter(all_row_group_indices),
+                   [](auto const& file_meta) {
+                     std::vector<size_type> rg_idx(file_meta.row_groups.size());
+                     std::iota(rg_idx.begin(), rg_idx.end(), 0);
+                     return rg_idx;
+                   });
+
+    // Set the current span of row group indices to the vector of all row group indices
+    current_row_group_indices = host_span<std::vector<size_type> const>(all_row_group_indices);
+  }
+  // Otherwise, set the current span of row group indices to the specified input row group indices
+  else {
+    current_row_group_indices = row_group_indices;
+  }
+
+  // Compute total number of input row groups
+  size_type const total_input_row_groups = row_group_indices.empty()
+                                             ? num_row_groups  // Use pre-computed member variable
+                                             : compute_total_row_groups(current_row_group_indices);
+
+  // Flag to check if the row groups will be filtered using row bounds
+  bool const is_trimmed_row_groups =
+    row_group_indices.empty() and (rows_to_skip > 0 or rows_to_read < max_num_rows);
+
+  // Use row bounds to filter row group indices if possible
+  std::vector<std::vector<size_type>> trimmed_row_group_indices;
+  std::vector<std::vector<size_type>> row_group_num_rows;
+  std::vector<std::vector<size_type>> row_group_row_offsets;
+
+  if (is_trimmed_row_groups) {
+    std::tie(trimmed_row_group_indices, row_group_num_rows, row_group_row_offsets) =
+      apply_row_bounds_filter(current_row_group_indices, rows_to_skip, rows_to_read);
+
+    // Update the current span of row group indices
+    current_row_group_indices = host_span<std::vector<size_type> const>(trimmed_row_group_indices);
+  }
+
+  // Compute number of input row groups after row bounds filter
+  size_type const total_row_bounds_filtered_row_groups =
+    is_trimmed_row_groups ? compute_total_row_groups(current_row_group_indices)
+                          : total_input_row_groups;
+
+  // Struct to store the number of row groups after filtering using stats and bloom filters
   surviving_row_group_metrics num_row_groups_after_filters{};
```

Comment:
Future PR:

This struct may be updated such that it tracks the number of row groups filtered with row bounds as well. In that case, this function will return the total_input_row_groups instead of `total_row_bounds_filtered_row_groups` (considered as `total_row_groups` before this PR)